sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CSharpSymbolUsageCollector.cs

```diff
@@ -89,13 +100,95 @@ public override void VisitAttribute(AttributeSyntax node)
+                case ExpressionStatementSyntax statement:
+                    // node;
+                    return SymbolAccess.None;
+                case InvocationExpressionSyntax invocation:
+                    // node(_) : <unexpected>
+                    return node == invocation.Expression ? SymbolAccess.Read : SymbolAccess.None;
+                case MemberAccessExpressionSyntax memberAccess:
+                    // _.node : node._
+                    return node == memberAccess.Name ? ParentAccessType(memberAccess) : SymbolAccess.Read;
+                case MemberBindingExpressionSyntax memberBinding:
+                    // _?.node : node?._
+                    return node == memberBinding.Name ? ParentAccessType(memberBinding) : SymbolAccess.Read;
+                case AssignmentExpressionSyntax assignmentExpression:
+                    // Ignoring distinction assignmentExpression.IsKind(SyntaxKind.SimpleAssignmentExpression) between
+                    // "node = _" and "node += _" both are considered as Write and rely on the parent to know if its read.
+                    //  node = _ : _ = node
+                    return node == assignmentExpression.Left ? SymbolAccess.Write | ParentAccessType(assignmentExpression) : SymbolAccess.Read;
+                case ArgumentSyntax argument:
+                    if (argument.RefOrOutKeyword.IsKind(SyntaxKind.OutKeyword))
+                    {
+                        //  out Type node : out node
+                        return SymbolAccess.Write | (argument.Expression.IsKind(SyntaxKindEx.DeclarationExpression) ? SymbolAccess.Declaration : SymbolAccess.None);
+                    }
+                    else if (argument.RefOrOutKeyword.IsKind(SyntaxKind.RefKeyword))
+                    {
+                        //  ref node
+                        return SymbolAccess.ReadWrite;
+                    }
+                    else
+                    {
+                        // nameof(node) : node
+                        return argument.Expression.IsInNameofCall(this.getSemanticModel(argument.Expression)) ? SymbolAccess.None : SymbolAccess.Read;
+                    }
+                case ExpressionSyntax expressionSyntax when expressionSyntax.IsAnyKind(IncrementKinds):
+                    // node++
+                    return SymbolAccess.Write | ParentAccessType(expressionSyntax);
+                case ArrowExpressionClauseSyntax arrowExpressionClause when arrowExpressionClause.Parent is MethodDeclarationSyntax arrowMethod:
+                    return arrowMethod.ReturnType != null && arrowMethod.ReturnType.IsKnownType(KnownType.Void, this.getSemanticModel(arrowMethod))
+                        ? SymbolAccess.None
+                        : SymbolAccess.Read;
+                default:
+                    // in case of SyntaxKind.SimpleLambdaExpression|ParenthesizedLambdaExpression|AnonymousMethodExpression with a void return type
```

Comment:
@alban-auzeill - for lambas and methods that actually have a body (i.e. they're not ArrowExpressions like above) - I am not sure if it's relevant to see if they return void - if there's multiple statements in the block, some may read, some may write